substrate/frame/bounties/src/lib.rs

```diff
@@ -873,6 +1037,124 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		// TODO: retries payment for funding the bounty or closing the bounty. updates the bounty
+		// status or removed it (when payment successful and bounty is closed).
+		#[pallet::call_index(10)]
+		// TODO: change weight
+		#[pallet::weight(<T as Config<I>>::WeightInfo::approve_bounty_with_curator())]
+		pub fn process_payment(
+			origin: OriginFor<T>,
+			#[pallet::compact] bounty_id: BountyIndex,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			Bounties::<T, I>::try_mutate_exists(
+				bounty_id,
+				|maybe_bounty| -> DispatchResultWithPostInfo {
+					let bounty = maybe_bounty.as_mut().ok_or(Error::<T, I>::InvalidIndex)?;
+					let payment_status = match bounty.status {
+						BountyStatus::Approved { ref mut payment_status } => payment_status,
+						BountyStatus::ApprovedWithCurator { ref mut payment_status, .. } =>
+							payment_status,
+						_ => return Err(Error::<T, I>::UnexpectedStatus.into()),
+					};
+
+					match payment_status {
+						PaymentState::Failed | PaymentState::Pending => {
+							// attempt payment again from failed/pending status
+							let bounty_account = Self::bounty_account_id(bounty_id);
+							let id = T::Paymaster::pay(
+								&bounty_account,
+								bounty.asset_kind.clone(),
+								bounty.value,
+							)
+							.map_err(|_| Error::<T, I>::FundingError)?;
+							*payment_status = PaymentState::Attempted { id };
+							return Ok(Pays::No.into());
```

Comment:
the caller should still pay. but you sill needs `DispatchResultWithPostInfo` type because this call in most cases will have bigger worth case weight than actual. the payout have two payments to process, the funding and the cancelation only one. I would process all of them with the same calls, `process_payment` and `check_payment_status`.